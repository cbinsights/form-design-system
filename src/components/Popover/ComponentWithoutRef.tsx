import React from 'react';

const ComponentWithoutRef = ({
  children,
}: {
  children: React.ReactNode;
}): JSX.Element => {
  return <a>{children}</a>;
};

export default ComponentWithoutRef;
