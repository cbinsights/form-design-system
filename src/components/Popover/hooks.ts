import { RefObject, useEffect } from 'react';

let overflowStyle: string | null = null;

/**
 * Hook that disables scroll on a DOM node when `isDisabled` is true.
 *
 * @param {Object} disableScrollRef - react ref to DOM node
 * @param {Boolean} isDisabled
 */
export const useDisableScroll = (
  disableScrollRef: RefObject<HTMLElement> | undefined,
  isDisabled: boolean
): void => {
  useEffect(() => {
    if (disableScrollRef?.current) {
      const domNode = disableScrollRef.current;
      if (isDisabled) {
        overflowStyle = domNode.style.overflow;
        domNode.style.overflow = 'hidden';
      } else if (overflowStyle !== null) {
        domNode.style.overflow = overflowStyle;
      }
    }
  }, [disableScrollRef, isDisabled]);
};

/**
 * Hook that invokes `closeCallback` if a popover is active when a user scrolls.
 *
 * @param {Object} closeOnScrollRef - react ref to scrolling DOM node
 * @param {Boolean} isActive - if the popover is currently open/active
 * @param {Boolean} closeCallback - function that closes the popover
 */
export const useCloseOnScroll = (
  closeOnScrollRef: RefObject<HTMLElement> | undefined,
  isActive: boolean,
  closeCallback: () => void
): void => {
  useEffect(() => {
    if (closeOnScrollRef?.current && isActive) {
      const scrollRef = closeOnScrollRef.current;
      scrollRef.addEventListener('scroll', function scrollLogic() {
        closeCallback();
        scrollRef.removeEventListener('scroll', scrollLogic);
      });
    }
  }, [closeOnScrollRef, isActive]);
};
