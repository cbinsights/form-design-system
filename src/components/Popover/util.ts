import { Placement } from '@popperjs/core';
import { Alignment, PopperAlignment, Position } from '.';

/**
 * https://popper.js.org/popper-documentation.html#Popper.placements
 *
 * @param {String} position prop value
 * @param {String} alignment prop value
 * @returns {String} valid `placement` value for PopperJS `Popper` component
 */
export const getPopperPlacement = (
  position: Position,
  alignment: Alignment
): Placement => {
  // Popperjs does not have a "center" placement variation. It centers by default.
  // Our component explicitly accepts a `center` value for alignment.
  const variation: '' | `-${PopperAlignment}` =
    !alignment || alignment === 'center' ? '' : `-${alignment}`;
  return `${position}${variation}`;
};
