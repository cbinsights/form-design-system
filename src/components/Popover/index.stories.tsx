import React, { useRef } from 'react';
import State from 'util/storybook-docs/State';
import Button from 'components/Button';
import Popover, { PopoverProps } from '.';

const Template = (args: PopoverProps): JSX.Element => <Popover {...args} />;
export const Primary = Template.bind({});
Primary.args = {
  transitionName: 'GrowFast',
  trigger: (
    <div className="margin--bottom">
      <Button label="Click Me" />
    </div>
  ),
  children: (
    <div
      className="padding--all"
      style={{ outline: '3px dotted red', background: '#FFFFFF' }}
    >
      popover content
    </div>
  ),
};

export const StyledPopoverContent = Template.bind({});
StyledPopoverContent.args = {
  trigger: <Button theme="outlined" label="Open popover" />,
  position: 'bottom',
  alignment: 'start',
  distance: 8,
  children: (
    <div className="bgColor--white rounded--all elevation--2 padding--all">
      <h3 className="type--head3">Look at me</h3>
      <p>
        <em>i am the popover</em>
      </p>
    </div>
  ),
};

export const ControlledPopover = (args: PopoverProps): JSX.Element => (
  <State
    initialValue={true}
    render={(myIsOpen, setMyIsOpen) => (
      <Popover
        {...args}
        trigger={
          <button
            onClick={() => {
              setMyIsOpen(!myIsOpen);
            }}
          >
            Open popover
          </button>
        }
        position="bottom"
        alignment="start"
        interactionMode="controlled"
        isOpen={myIsOpen}
        onUserDismiss={() => {
          setMyIsOpen(false);
        }}
        distance={8}
      >
        <div className="bgColor--white rounded--all elevation--2 padding--all">
          <h3 className="type--head3">Look at me</h3>
          <p>
            <em>i am the popover</em>
          </p>
        </div>
      </Popover>
    )}
  />
);

ControlledPopover.parameters = {
  docs: {
    description: {
      story:
        'The "controlled" `interactionMode` follows the conventions of a [controlled component](https://reactjs.org/docs/forms.html#controlled-components), meaning the open state of the popover is exclusively controlled by the prop `isOpen`. To update the state of `isOpen` based on user interaction, you must use the `onUserDismiss` callback. The `onUserDismiss` function is invoked when 1. The user clicks the trigger element 2. The user presses `Escape` 3. the user clicks away from the popover. If you\'d like to disable scrolling of some element when a popover is open you can pass a ref of the scrolling element to the `disableScrollRef` prop.',
    },
  },
};

export const PopoverWithinAPopover = (args: PopoverProps): JSX.Element => (
  <Popover trigger={<button>Open popover</button>}>
    <div className="bgColor--white rounded--all elevation--2 padding--all">
      <div className="padding--y--s padding--x--m" style={{ overflow: 'hidden' }}>
        <h3>Here is some content inside.</h3>
        <p>This popover is overflow hidden.</p>
        <p>
          Here is something else to
          <Popover
            {...args}
            trigger={<button className="margin--left--xs">click</button>}
          >
            <div className="bgColor--white rounded--all elevation--2 padding--all">
              <div className="padding--all--m">
                <p>
                  You can see this modal even though the parent modal has its{' '}
                  <code>overflow</code> set to <code>hidden</code>.
                </p>
                <p>You can also click inside here without closing the parent modals.</p>
              </div>
            </div>
          </Popover>
        </p>
      </div>
    </div>
  </Popover>
);
PopoverWithinAPopover.args = {
  isFixed: true,
  disablePortal: 'true',
};
const hidden = { table: { disable: true } };
PopoverWithinAPopover.argTypes = {
  triggerElement: hidden,
  theme: hidden,
  children: hidden,
  placement: hidden,
  isOpen: hidden,
  onUserDismiss: hidden,
  triggerType: hidden,
};

export const CloseOnScroll = (args: PopoverProps): JSX.Element => {
  const containerEl = useRef(null);
  return (
    <div
      ref={containerEl}
      style={{ height: 200, border: '1px dashed red', overflow: 'scroll' }}
    >
      <div style={{ height: 500, background: 'yellow' }}>
        <Popover
          {...args}
          trigger={<Button theme="outlined" label="Open popover" />}
          position="bottom"
          alignment="start"
          distance={8}
          closeOnScrollRef={containerEl}
        >
          <div className="bgColor--white rounded--all elevation--2 padding--all">
            <h3 className="type--head3">Look at me</h3>
            <p>
              <em>i am the popover</em>
            </p>
          </div>
        </Popover>
      </div>
    </div>
  );
};

CloseOnScroll.parameters = {
  docs: {
    description: {
      story:
        "If you'd like the ability to close the popover once the user scrolls some element, you can pass a ref of that element to the `closeOnScrollRef` prop. If you're not sure how to handle scrolling behavior with a popover, choose this option.",
    },
  },
};

export const DisableScrolling = (args: PopoverProps): JSX.Element => {
  const containerEl = useRef(null);
  return (
    <div
      ref={containerEl}
      style={{ height: 200, border: '1px dashed red', overflow: 'scroll' }}
    >
      <div style={{ height: 500, background: 'yellow' }}>
        <Popover
          {...args}
          trigger={<Button theme="outlined" label="Open popover" />}
          position="bottom"
          alignment="start"
          distance={8}
          disableScrollRef={containerEl}
        >
          <div className="bgColor--white rounded--all elevation--2 padding--all">
            <h3 className="type--head3">Look at me</h3>
            <p>
              <em>i am the popover</em>
            </p>
          </div>
        </Popover>
      </div>
    </div>
  );
};

DisableScrolling.parameters = {
  docs: {
    description: {
      story:
        "If you'd like to disable scrolling of some element when a popover is open you can pass a ref of the scrolling element to the `disableScrollRef` prop.",
    },
  },
};

export default {
  component: Popover,
  title: 'components/Popover',
  parameters: {
    docs: {
      description: {
        component:
          'Base popover component. Bring your own trigger and popover content. What this component does: Handle events according to `interactionMode` passed in & handle positioning of popover content What this component does **not** do: Content styling (the popover container is just a transparent div) & Arrows. They arent supported, but we can add them in the future. When setting `position` and `alignment` props, the positioning engine will prefer the props you provide, but may move the popover content to avoid edges of the viewport. This behavior is intentional.',
      },
    },
  },
};
