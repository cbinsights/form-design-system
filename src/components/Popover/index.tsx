import React, { useEffect, useState, RefObject, HTMLAttributes, useRef } from 'react';
import cc from 'classcat';
import * as RadixPopover from '@radix-ui/react-popover';
import { CSSTransition } from 'react-transition-group';
import FDS from 'dictionary/js/styleConstants';
import { useDisableScroll, useCloseOnScroll } from './hooks';

export type Position = 'top' | 'right' | 'bottom' | 'left';
export type InteractionMode = 'hover' | 'click' | 'controlled';

// gets non-optional function type because Parameters doesnt take `undefined`
type RequiredOnInteractOutsideFn = Required<
  React.ComponentProps<typeof RadixPopover.Content>
>['onInteractOutside'];

// getting the type of the event parameter to use in onUserDismiss from the tuple returned by Parameters
type InteractOutsideEvent = Parameters<RequiredOnInteractOutsideFn>[0];
export interface PopoverProps {
  /** JSX - Acts as a positioning reference for the popover and triggers active state */
  trigger: React.ReactElement;

  /**
   * JSX - Content to place in the positioned popover container. The popover container
   * does not provide any default styling; content should be styled with background,
   * borders, and shadows as appropriate.
   */
  children: React.ReactNode;

  /**
   * What mode of interaction triggers the active/inactive state of the popover.
   *
   * `hover` - popover opens on trigger hover/focus
   *
   * `click` - popover opens on trigger click
   *
   * `controlled` - enables "fully controlled" mode in which the popover is only active
   * when the `isOpen` prop is set to `true`. Use `onUserDismiss` to handle dismissal
   * events from user interaction.
   */
  interactionMode?: InteractionMode;

  /** Controlls active state of popover when in fully controlled interaction mode */
  isOpen?: boolean;

  /**
   * Callback fired when user takes an action to dismiss the popover.
   * (e.g. ESC press, clicking outside, etc.)
   * Useful for updating `isOpen` in controlled mode.
   * Function is invoked with the DOM event as the argument.
   */
  onUserDismiss?: (e: MouseEvent | KeyboardEvent | InteractOutsideEvent) => void;

  /** disables portaling the popover to `document.body` */
  disablePortal?: boolean;

  /**
   * Vertical position preference of popover content.
   * `top` for example, will place the popover content above the trigger.
   */
  position?: Position;

  /** Horizontal alignment preference of popover content relative to trigger */
  alignment?: 'start' | 'center' | 'end';

  /** Offset distance from trigger. */
  distance?: number;

  /** Delay in milliseconds for popover to trigger (hover mode only) */
  delay?: number;

  /** Name of transition for popover content */
  transitionName?: 'GrowFast';

  /** Callback called when popover opens */
  onOpen?: () => void;

  /** Callback called when popover closes */
  onClose?: () => void;

  /** When popover opens, this element should freeze scrolling. When the popover closes, the scrollability of this element should resume. */
  disableScrollRef?: RefObject<HTMLElement>;

  /** When the Popover detects scroll events from this ref, the popover should close. */
  closeOnScrollRef?: RefObject<HTMLElement>;

  /** Ref to forward to the content container of the popover */
  ref?: RefObject<HTMLDivElement>;
}

const Popover = React.forwardRef<HTMLDivElement, PopoverProps>(
  (
    {
      alignment = 'start',
      children,
      closeOnScrollRef,
      delay = 0,
      disablePortal,
      disableScrollRef,
      distance = 4,
      interactionMode = 'click',
      isOpen = false,
      onClose = () => {},
      onOpen = () => {},
      onUserDismiss = () => {},
      position = 'bottom',
      transitionName,
      trigger,
    }: PopoverProps,
    forwardedRef
  ) => {
    const [isActive, setIsActive] = useState(isOpen);
    const refContent = forwardedRef || useRef<HTMLDivElement>(null);
    useCloseOnScroll(closeOnScrollRef, isActive, () => {
      handleChange(false);
    });
    useDisableScroll(disableScrollRef, isActive);

    const isHover = interactionMode === 'hover';
    const isControlled = interactionMode === 'controlled';

    // update active state on props change to accommodate fully controlled popovers
    useEffect(() => {
      setIsActive(isControlled && !!isOpen);
    }, [interactionMode, isOpen]);

    const triggerProps: HTMLAttributes<HTMLElement> = {
      style: isActive && isControlled ? { pointerEvents: 'none' } : {},
    };
    let hoverTimeout: NodeJS.Timeout;
    if (interactionMode === 'hover') {
      triggerProps.onMouseEnter = () => {
        if (delay > 0) {
          hoverTimeout = setTimeout(handleChange, delay, true);
        } else {
          handleChange(true);
        }
      };
      triggerProps.onMouseLeave = () => {
        if (hoverTimeout) {
          clearTimeout(hoverTimeout);
        }
        handleChange(false);
      };
      triggerProps.onClick = (e) => {
        e.stopPropagation();
        handleChange(false);
      };
      triggerProps.tabIndex = 1;
    } else if (interactionMode === 'click') {
      triggerProps.onClick = (e) => {
        e.stopPropagation();
        handleChange(!isActive);
      };
    }

    const clonedTrigger = React.cloneElement(trigger, triggerProps);

    const contentStyle = {
      zIndex: Number(FDS.ZINDEX_MODAL),
    };

    const popperContent = (
      <CSSTransition
        in={isActive}
        unmountOnExit={isHover}
        timeout={transitionName ? 200 : 0}
        classNames={transitionName && !isActive ? `rtg${transitionName}` : undefined}
      >
        <div
          style={{
            ...contentStyle,
          }}
        >
          {children}
        </div>
      </CSSTransition>
    );

    const handleChange = (open: boolean) => {
      if (!isControlled) {
        setIsActive(open);
        if (open) {
          onOpen();
        } else {
          onClose();
        }
      }
    };

    return (
      <RadixPopover.Root open={isActive} onOpenChange={handleChange}>
        <RadixPopover.Trigger asChild>
          <div className="display--inlineFlex div--100">{clonedTrigger}</div>
        </RadixPopover.Trigger>
        <RadixPopover.Content
          ref={refContent}
          className={cc([{ 'ease-in-out': !closeOnScrollRef }])}
          align={alignment}
          side={position}
          sideOffset={distance}
          portalled={!disablePortal}
          onEscapeKeyDown={onUserDismiss}
          onInteractOutside={onUserDismiss}
          avoidCollisions
          collisionTolerance={0}
        >
          {popperContent}
        </RadixPopover.Content>
      </RadixPopover.Root>
    );
  }
);

Popover.displayName = 'Popover';

export default Popover;
