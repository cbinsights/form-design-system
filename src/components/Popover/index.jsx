import React, { useRef, useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { Manager, Reference, Popper } from 'react-popper';
import { CSSTransition } from 'react-transition-group';
import FDS from 'lib/dictionary/js/styleConstants';
import { isNotRefsEvent } from 'components/util/events';

export const VALID_POSITIONS = ['auto', 'top', 'right', 'bottom', 'left'];
export const VALID_ALIGNMENTS = ['start', 'end', 'center'];
export const VALID_INTERACTION_MODES = ['hover', 'click', 'controlled'];

/**
 * https://popper.js.org/popper-documentation.html#Popper.placements
 *
 * @param {String} position prop value
 * @param {String} alignment prop value
 * @returns {String} valid `placement` value for PopperJS `Popper` component
 */
export const getPopperPlacement = (position, alignment) => {
  // Popperjs does not have a "center" placement variation. It centers by default.
  // Our component explicitly accepts a `center` value for alignment.
  const variation = !alignment || alignment === 'center' ? '' : `-${alignment}`;
  return `${position}${variation}`;
};

let overflowStyle = null;

/**
 * Hook that disables scroll on a DOM node when `isDisabled` is true.
 *
 * @param {Object} disableScrollRef - react ref to DOM node
 * @param {Boolean} isDisabled
 */
const useDisableScroll = (disableScrollRef, isDisabled) => {
  useEffect(() => {
    if (disableScrollRef) {
      const domNode = disableScrollRef.current || disableScrollRef;
      if (isDisabled) {
        overflowStyle = domNode.style.overflow;
        domNode.style.overflow = 'hidden';
      } else if (overflowStyle !== null) {
        domNode.style.overflow = overflowStyle;
      }
    }
  }, [disableScrollRef, isDisabled]);
};

/**
 * Hook that invokes `closeCallback` if a popover is active when a user scrolls.
 *
 * @param {Object} closeOnScrollRef - react ref to scrolling DOM node
 * @param {Boolean} isActive - if the popover is currently open/active
 * @param {Boolean} closeCallback - function that closes the popover
 */
const useCloseOnScroll = (closeOnScrollRef, isActive, closeCallback) => {
  useEffect(() => {
    if (closeOnScrollRef && isActive) {
      const scrollRef = closeOnScrollRef.current || closeOnScrollRef;
      scrollRef.addEventListener('scroll', function scrollLogic() {
        closeCallback();
        scrollRef.removeEventListener('scroll', scrollLogic);
      });
    }
  }, [closeOnScrollRef, isActive]);
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Popover = React.forwardRef(
  (
    {
      interactionMode = 'click',
      disablePortal,
      position = 'auto',
      alignment = 'start',
      distance = 4,
      delay = 0,
      onOpen = () => {},
      onClose = () => {},
      trigger,
      children,
      isOpen,
      transitionName,
      disableScrollRef,
      closeOnScrollRef,
    },
    forwardedRef
  ) => {
    const [isActive, setIsActive] = useState(false);
    const refTriggerWrap = useRef(null);
    const refContent = forwardedRef || React.createRef();

    const handleSetIsActive = (newValue) => {
      setIsActive(() => {
        if (newValue) {
          onOpen();
        } else {
          onClose();
        }
        return newValue;
      });
    };

    useCloseOnScroll(closeOnScrollRef, isActive, () => handleSetIsActive(false));
    useDisableScroll(disableScrollRef, isActive);

    // update active state on props change to accommodate fully controlled popovers
    useEffect(() => {
      handleSetIsActive(interactionMode === 'controlled' && isOpen);
    }, [interactionMode, isOpen]);

    /**
     * Closes popover when user presses ESC
     * @param {Event} e DOMEvent
     */
    const handleKeyPress = (e) => {
      const isEscapeKey = ['Esc', 'Escape'].some((key) => key === e.key);
      if (isEscapeKey) handleSetIsActive(false);
    };

    /**
     * Closes popover when user clicks outside of content or trigger
     * @param {Event} e DOMEvent
     */
    const handleBodyClick = (e) => {
      const isNotPopoverClick = isNotRefsEvent([refTriggerWrap, refContent], e);
      if (isNotPopoverClick) handleSetIsActive(false);
    };

    useEffect(() => {
      /* eslint-disable no-undef */
      document.body.addEventListener('mousedown', handleBodyClick, false);
      document.body.addEventListener('keyup', handleKeyPress, false);

      return () => {
        document.body.removeEventListener('mousedown', handleBodyClick, false);
        document.body.removeEventListener('keyup', handleKeyPress, false);
      };
      /* eslint-enable no-undef */
    });

    let triggerProps = {};
    let hoverTimeout;
    switch (interactionMode) {
      case 'hover':
        triggerProps.onMouseEnter = () => {
          if (parseInt(delay, 10) > 0) {
            hoverTimeout = setTimeout(handleSetIsActive, delay, true);
          } else {
            handleSetIsActive(true);
          }
        };
        triggerProps.onMouseLeave = () => {
          if (hoverTimeout) {
            clearTimeout(hoverTimeout);
          }
          handleSetIsActive(false);
        };
        triggerProps.onKeyUp = (e) => {
          if (e.key === 'Tab') {
            handleSetIsActive(true);
          }
        };
        triggerProps.onKeyDown = (e) => {
          if (e.key === 'Tab') {
            handleSetIsActive(false);
          }
        };
        triggerProps.tabIndex = '1';
        break;
      case 'click':
        triggerProps.onClick = () => {
          handleSetIsActive(!isActive);
        };
        break;
      default:
        triggerProps = {};
    }

    const clonedTrigger = React.cloneElement(trigger, triggerProps);

    // https://popper.js.org/popper-documentation.html#modifiers
    const popperModifiers = [
      {
        name: 'offset',
        enabled: true,
        options: {
          offset: [0, distance],
        },
      },
      {
        name: 'preventOverflow',
        options: {
          boundariesElement: 'viewport',
        },
      },
      {
        name: 'computeStyles',
        options: {
          // When gpuAcceleration is enabled, `react-popper` always
          // positions content to top/left 0 and uses `translate3d` to
          // nudge it into place.
          //
          // This can cause issues with CSS transitions because it sets
          // `will-change: transform`, which forces transitions to also
          // ease the `translate3d`, which causes an unintended
          // "diagonal slide" effect as the content appears.
          gpuAcceleration: false,
        },
      },
    ];

    const contentStyle = {
      zIndex: FDS.ZINDEX_MODAL,
    };

    const popperContent = (
      <CSSTransition
        in={isActive}
        unmountOnExit
        timeout={transitionName ? 200 : 0}
        classNames={transitionName ? `rtg${transitionName}` : undefined}
      >
        <Popper
          innerRef={(node) => {
            refContent.current = node;
          }}
          modifiers={popperModifiers}
          placement={getPopperPlacement(position, alignment)}
        >
          {({ placement, ref, style }) => (
            <div
              ref={ref}
              style={{
                ...contentStyle,
                ...style,
              }}
              data-placement={placement}
            >
              {children}
            </div>
          )}
        </Popper>
      </CSSTransition>
    );

    return (
      <Manager>
        <Reference>
          {({ ref }) => (
            <div ref={ref} aria-haspopup="true" aria-expanded={isActive.toString()}>
              <div ref={refTriggerWrap}>{clonedTrigger}</div>
            </div>
          )}
        </Reference>
        {disablePortal
          ? popperContent
          : ReactDOM.createPortal(
              popperContent,
              document.body /* eslint-disable-line no-undef */
            )}
      </Manager>
    );
  }
);

Popover.displayName = 'Popover';

const validRef = PropTypes.oneOfType([
  PropTypes.func,
  PropTypes.shape({ current: PropTypes.any }),
]);

Popover.propTypes = {
  /** JSX - Acts as a positioning reference for the popover and triggers active state */
  trigger: PropTypes.oneOfType([PropTypes.node, PropTypes.element]).isRequired,

  /**
   * JSX - Content to place in the positioned popover container. The popover container
   * does not provide any default styling; content should be styled with background,
   * borders, and shadows as appropriate.
   */
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.element]).isRequired,

  /**
   * What mode of interaction triggers the active/inactive state of the popover.
   *
   * `hover` - popover opens on trigger hover/focus
   *
   * `click` - popover opens on trigger click
   *
   * `controlled` - enables "fully controlled" mode in which the popover is only active
   * when the `isOpen` prop is set to `true`
   */
  interactionMode: PropTypes.oneOf(VALID_INTERACTION_MODES),

  /** Controlls active state of popover when in fully controlled interaction mode */
  isOpen: PropTypes.bool,

  /** disables portaling the popover to `document.body` */
  disablePortal: PropTypes.bool,

  /**
   * Vertical position preference of popover content.
   * `top` for example, will place the popover content above the trigger.
   */
  position: PropTypes.oneOf(VALID_POSITIONS),

  /** Horizontal alignment preference of popover content relative to trigger */
  alignment: PropTypes.oneOf(VALID_ALIGNMENTS),

  /** Offset distance from trigger. */
  distance: PropTypes.number,

  /** Delay in milliseconds for popover to trigger (hover mode only) */
  delay: PropTypes.number,

  /** Name of transition for popover content */
  transitionName: PropTypes.oneOf(['GrowFast']),

  /** Callback called when popover opens */
  onOpen: PropTypes.func,

  /** Callback called when popover closes */
  onClose: PropTypes.func,

  /** When popover opens, this element should freeze scrolling. When the popover closes, the scrollability of this element should resume. */
  disableScrollRef: validRef,

  /** When the Popover detects scroll events from this ref, the popover should close. */
  closeOnScrollRef: validRef,

  /** Ref to forward to the content container of the popover */
  ref: validRef,
};

export default Popover;
