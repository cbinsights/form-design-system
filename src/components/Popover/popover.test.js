import React from 'react';
import { shallow, mount } from 'enzyme';

import Popover, { getPopperPlacement } from '.';

const Content = () => (<p>popover content</p>);
const triggerJsx = (<button>trigger</button>);

const SELECTOR_TRIGGER = '[aria-haspopup] button';

/**
 * @param {ReactWrapper} w enzyme react wrapper
 * @returns {Boolean} if the CSSTransition `in` prop is allowing the content to display
 */
const isPopperOpen = (w) => w.find('CSSTransition').prop('in');

describe('Popover component', () => {

  it('matches snapshot (default props)', () => {
    const wrapper = mount(<Popover trigger={triggerJsx}><Content /></Popover>);
    expect(wrapper).toMatchSnapshot();
  });

  it('uses Portal by default', () => {
    const wrapper = mount(<Popover trigger={triggerJsx}><Content /></Popover>);
    expect(wrapper.exists('Portal')).toBe(true);
  });

  it('does NOT use Portal when disablePortal is set', () => {
    const wrapper = mount(<Popover trigger={triggerJsx} disablePortal><Content /></Popover>);
    expect(wrapper.exists('Portal')).toBe(false);
  });

  it('ignores `isOpen` prop when not in controlled mode', () => {
    const wrapper = shallow(
      <Popover
        trigger={triggerJsx}
        isOpen
      >
        <Content />
      </Popover>
    );
    expect(isPopperOpen(wrapper)).toBe(false);
  });

  describe('getPopperPlacement', () => {
    it('returns an alignment in the placement', () => {
      const actual = getPopperPlacement('bottom', 'start');
      const expected = 'bottom-start';
      expect(actual).toBe(expected);
    });
    it('does not return alignment when none is passed in', () => {
      const actual = getPopperPlacement('bottom');
      const expected = 'bottom';
      expect(actual).toBe(expected);
    });
    it('does not return placement alignment when "center" alignment is passed', () => {
      const actual = getPopperPlacement('bottom', 'center');
      const expected = 'bottom';
      expect(actual).toBe(expected);
    });
  });

  describe('interactionMode="click"', () => {
    let wrapper;
    let triggerEl;
    let onOpen;
    let onClose;

    beforeEach(() => {
      onOpen = jest.fn()
      onClose = jest.fn()
      wrapper = mount(
        <Popover trigger={triggerJsx} interactionMode="click" onOpen={onOpen} onClose={onClose} >
          <Content />
        </Popover>
      );
      triggerEl = wrapper.find(SELECTOR_TRIGGER);
    });

    afterEach(() => {
      wrapper = null;
      triggerEl = null;
    });

    it('adds click-related props to trigger', () => {
      const triggerPropNames = Object.keys(triggerEl.props());
      expect(triggerPropNames).toContain('onClick');
    });

    it('opens and closes popover on trigger clicks', () => {
      expect(isPopperOpen(wrapper)).toBe(false);

      triggerEl.simulate('click');
      expect(isPopperOpen(wrapper)).toBe(true);
      expect(onOpen).toHaveBeenCalled()

      triggerEl.simulate('click');
      expect(isPopperOpen(wrapper)).toBe(false);
      expect(onClose).toHaveBeenCalled()
    });

    it('closes on clicks away from popover', () => {
      expect(isPopperOpen(wrapper)).toBe(false);

      triggerEl.simulate('click');
      expect(isPopperOpen(wrapper)).toBe(true);
      expect(onOpen).toHaveBeenCalled()

      triggerEl.simulate('click', { target: '<p>lol not the popover</p>' });
      expect(isPopperOpen(wrapper)).toBe(false);
      expect(onClose).toHaveBeenCalled()
    });
  });

  describe('interactionMode="hover"', () => {
    let wrapper;
    let triggerEl;
    let onOpen;
    let onClose;

    beforeEach(() => {
      onOpen = jest.fn()
      onClose = jest.fn()
      wrapper = mount(
        <Popover trigger={triggerJsx} interactionMode="hover"  onOpen={onOpen} onClose={onClose} >
          <Content />
        </Popover>
      );
      triggerEl = wrapper.find(SELECTOR_TRIGGER);
    });

    afterEach(() => {
      wrapper = null;
      triggerEl = null;
    });

    it('adds hover-related props to trigger', () => {
      const triggerPropNames = Object.keys(triggerEl.props());
      ['onMouseEnter', 'onKeyUp', 'onKeyDown', 'tabIndex'].forEach((prop) => {
        expect(triggerPropNames).toContain(prop);
      });
    });

    it('opens and closes popover on keyboard navigating in and out via Tab', () => {
      expect(isPopperOpen(wrapper)).toBe(false);

      triggerEl.simulate('keyup', { key: 'Tab' });
      expect(isPopperOpen(wrapper)).toBe(true);
      expect(onOpen).toHaveBeenCalled()

      triggerEl.simulate('keydown', { key: 'Tab' });
      expect(isPopperOpen(wrapper)).toBe(false);
      expect(onClose).toHaveBeenCalled()
    });

    it('closes on mouseleave outside of Popover elements', () => {
      expect(isPopperOpen(wrapper)).toBe(false);

      triggerEl.simulate('mouseenter');
      expect(isPopperOpen(wrapper)).toBe(true);
      expect(onOpen).toHaveBeenCalled()

      triggerEl.simulate('mouseleave', { target: '<p>lol not the popover</p>' });
      expect(isPopperOpen(wrapper)).toBe(false);
      expect(onClose).toHaveBeenCalled()
    });

  });

});
