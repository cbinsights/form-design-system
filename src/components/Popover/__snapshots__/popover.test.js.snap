// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover component matches snapshot (default props) 1`] = `
<Popover
  trigger={
    <button>
      trigger
    </button>
  }
>
  <Manager>
    <Reference>
      <div
        aria-expanded="false"
        aria-haspopup="true"
      >
        <div>
          <button
            onClick={[Function]}
          >
            trigger
          </button>
        </div>
      </div>
    </Reference>
    <Portal
      containerInfo={<body />}
    >
      <CSSTransition
        classNames=""
        in={false}
        timeout={0}
        unmountOnExit={true}
      >
        <Transition
          appear={false}
          enter={true}
          exit={true}
          in={false}
          mountOnEnter={false}
          onEnter={[Function]}
          onEntered={[Function]}
          onEntering={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onExiting={[Function]}
          timeout={0}
          unmountOnExit={true}
        />
      </CSSTransition>
    </Portal>
  </Manager>
</Popover>
`;
